flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImplTest.java

```diff
@@ -0,0 +1,371 @@
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.collection.IsEmptyCollection.empty;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Tests for the {@link JobAwareShuffleEnvironmentImpl}.
+ */
+public class JobAwareShuffleEnvironmentImplTest extends TestLogger {
+
+	private final JobID jobId = new JobID();
+	private TestResultPartitionWriter retainedOnConsumptionWriter;
+	private TestResultPartitionWriter releasedOnConsumptionWriter;
+	private TestShuffleEnvironment wrappedEnvironment;
+	private JobAwareShuffleEnvironmentImpl<?> jobAwareEnvironment;
+	private Collection<ResultPartitionWriter> writers;
+
+	@Before
+	public void setup() {
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		retainedOnConsumptionWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+		releasedOnConsumptionWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+		TestResultPartitionWriter releasedOnConsumptionWithoutLocalResourcesWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+		TestResultPartitionWriter retainedOnConsumptionWithoutLocalResourcesWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+
+		wrappedEnvironment = new TestShuffleEnvironment(
+			Arrays.asList(retainedOnConsumptionWriter, releasedOnConsumptionWriter));
+
+		jobAwareEnvironment = new JobAwareShuffleEnvironmentImpl<>(wrappedEnvironment);
+
+		ResultPartitionDeploymentDescriptor deploymentDescriptor1 = createResultPartitionDeploymentDescriptor(
+			retainedOnConsumptionWriter.resultPartitionID, false, true);
+		ResultPartitionDeploymentDescriptor deploymentDescriptor2 = createResultPartitionDeploymentDescriptor(
+			releasedOnConsumptionWriter.resultPartitionID, true, true);
+		ResultPartitionDeploymentDescriptor deploymentDescriptor3 = createResultPartitionDeploymentDescriptor(
+			releasedOnConsumptionWithoutLocalResourcesWriter.resultPartitionID, false, false);
+		ResultPartitionDeploymentDescriptor deploymentDescriptor4 = createResultPartitionDeploymentDescriptor(
+			retainedOnConsumptionWithoutLocalResourcesWriter.resultPartitionID, true, false);
+
+		writers = jobAwareEnvironment.createResultPartitionWriters(
+			jobId,
+			"test",
+			executionAttemptId,
+			Arrays.asList(deploymentDescriptor1, deploymentDescriptor2, deploymentDescriptor3, deploymentDescriptor4),
```

Comment:
Not sure whether we return for every `deploymentDescriptor` a `ResultPartitionWriter` from `wrappedEnvironment`. I guess it would be good idea to add a check state to the `JobAwareShuffleEnvironment` that as many writers need to come out of the `#createResultPartitionWriters` method as `ResultPartitionDeploymentDescriptors` are put in.

Reply:
The ShuffleEnvironment docs specify that the order of the returned partitions writers must correspond to the order of input deployment descriptors. One can interpret this to also mean that the number of writers must be equal to the number of deployment descriptors.
@azagrebin

Reply:
Then, the test implementation violates this constraint. Because we only return the two writers which we created above.